Use process warnings for deprecations

eritbh opened the pull request:
Addresses #1274. Adds process warning emits alongside existing deprecation-related client `warn` emits. To avoid emitting the same process warning multiple times, uses a wrapper around `process.emitWarning` that checks if the given warning code has already been used, and if so, skips emitting the warning again.

eritbh committed:
Create deprecation helper, use for existing warns

eritbh committed:
use correct Set method, derp

eritbh commented on lib/util/emitDeprecation.js:
```diff
@@ -0,0 +1,9 @@
+    }
+    emittedCodes.add(code);
+    process.emitWarning(description, "DeprecationWarning", code);
+};
```
Hopefully this matches what you meant by [this](https://github.com/abalabahaha/eris/discussions/1274#discussioncomment-1285642) @bsian03?

I've opted to use a `Set` because a quick search seemed to indicate `Set#has` is more performant than `Array#includes`, though the number of warnings we'll have is probably never going to be big enough for it to truly matter. Could use an array instead for simplicity.

bsian03 replied:
See my discussion comment

eritbh replied:
Performance comparison done and discussed on Discord, https://discord.com/channels/831967755447828491/884930349119991849/885222846815678495, arrays are faster. Addressed in e3313be.

eritbh commented on lib/Client.js:
```diff
@@ -286,6 +287,7 @@ class Client extends EventEmitter {
     */
     addMessageReaction(channelID, messageID, reaction, userID) {
         if(userID !== undefined) {
+            emitDeprecation("eris:1", "addMessageReaction() was called without an \"@me\" `userID` argument");
```
I've made up our warning codes completely arbitrarily, just numbering them sequentially as I went. We could use more descriptive codes (e.g. associating them with issues/PRs that implement or describe the deprecation) or we could use a method like this; whatever works, as long as it's a standard naming scheme across the project.

JustCat80 replied:
I feel like to make it harder to overlap in different prs we should use the pr/issue or an actual name or something 
so for example either "eris:#1239"(maybe not because multiple different warnings so maybe a suffix of the deprecation number within the pr or smthn) or "eris:embed" or similar for Support sending multiple embeds (#1239)

bsian03 replied:
Could probably discuss this in the discussion

eritbh replied:
Discussed in https://github.com/abalabahaha/eris/discussions/1274#discussioncomment-1293529 and done in 672183b

JustCat80 commented on lib/Client.js:
```diff
@@ -286,6 +287,7 @@ class Client extends EventEmitter {
     addMessageReaction(channelID, messageID, reaction, userID) {
         if(userID !== undefined) {
+            emitDeprecation("eris:1", "addMessageReaction() was called without an \"@me\" `userID` argument");
             this.emit("warn", "[DEPRECATED] addMessageReaction() was called without an \"@me\" `userID` argument");
```
Shouldnt these emit warns be removed? as the emitDeprecation should handle it no?
(github highlighted the wrong line or im dumb)

eritbh replied:
No, we're keeping them for now for backwards compatibility, but we can remove them in a future breaking change. See the summary I wrote in #1274.

JustCat80 replied:
Ah I see thats fair, kinda ironic that the deprecations are deprecated tho lol

eritbh committed:
Use array for emittedCodes

Array ops are apparently faster than Set ops in V8? https://discord.com/channels/831967755447828491/884930349119991849/885222846815678495

eritbh committed:
Move warning messages to a single location

eritbh committed:
Use named warning codes

See https://github.com/abalabahaha/eris/discussions/1274#discussioncomment-1293529

eritbh committed:
Alphabetize deprecation codes

eritbh committed:
Don't include `eris:` in every code manually

eritbh commented on lib/util/emitDeprecation.js:
```diff
@@ -0,0 +1,20 @@
+const warningMessages = {
+    CREATE_CHANNEL_OPTIONS: "createChannel() was called with a string `options` argument",
+    CREATE_CHANNEL_REASON: "createChannel() was called with a string `reason` argument",
```
These two messages are basically the same, and both have the same root cause - the signature of `createChannel` has changed. Should we merge them into the same code?

bsian03 replied:
Yup

eritbh replied:
Done in b3a75e4

eritbh reviewed (commented):
Changed structure around a bit in response to discussion activity. Looking for feedback on the codes themselves.

I also think it would be a good idea to rewrite some of these existing messages to include more context and provide migration paths away from the problem, rather than focusing on being short. How do we feel about this?

eritbh committed:
Merge + rewrite createChannel deprecations

eritbh committed:
rewrite all the other deprecation warnings

bsian03 commented on lib/util/emitDeprecation.js:
```diff
@@ -0,0 +1,19 @@
+        return;
+    }
+    emittedCodes.push(code);
+    process.emitWarning(warningMessages[code], "DeprecationWarning", `eris:${code}`);
```
Don't know how relevant this is since all of this is done in the backend but should invalid codes be ignored? Let's say someone accidentally pushes a typo

DonovanDMC replied:
what about like `warningMessages[code] || Invalid Warning Code "${code}"`, so that it doesn't just stay hidden

eritbh replied:
An invalid warning code would cause `warningMessages[code]` to be `undefined`, which results in a thrown error. Given this would come up immediately in testing any PR that adds new deprecation warnings, I don't think it's super important to explicitly handle this, but I can add it if you'd prefer the fallback behavior be more graceful anyway.

![image](https://user-images.githubusercontent.com/4165301/134074037-907266d3-1e53-4c0b-bc4c-c988bbf3b25b.png)

bsian03 replied:
Yes please implement a fallback

eritbh replied:
Done in bc83b95.

eritbh committed:
Include fallback message for invalid warning codes

eritbh committed:
Merge branch 'dev' into deprecation-process-warnings

eritbh committed:
Remove now-unused deprecation code

Single `embed` options were un-deprecated in dc72ad9

eritbh committed:
eslint

eritbh committed:
Merge branch 'abalabahaha:dev' into deprecation-process-warnings

eritbh committed:
Merge branch 'dev' into deprecation-process-warnings

eritbh committed:
Merge branch 'dev' into deprecation-process-warnings

eritbh committed:
Merge branch 'dev' into deprecation-process-warnings

DonovanDMC committed:
Merge branch 'dev' into deprecation-process-warnings

eritbh committed:
Merge branch 'dev' into deprecation-process-warnings

DonovanDMC committed:
Merge branch 'dev' into deprecation-process-warnings

eritbh committed:
Merge branch 'dev' into deprecation-process-warnings

eritbh committed:
Merge branch 'dev' into deprecation-process-warnings
